calyx-py/test/correctness/queues/fifo.py

```diff
@@ -23,20 +24,20 @@ def insert_fifo(prog, name, queue_len_factor=QUEUE_LEN_FACTOR):
     # If it is 0, we pop.
     # If it is 1, we peek.
     # If it is 2, we push `value` to the queue.
-    value = fifo.input("value", 32)  # The value to push to the queue
+    value = fifo.input("value", val_width)  # The value to push to the queue
 
     max_queue_len = 2**queue_len_factor
-    mem = fifo.seq_mem_d1("mem", 32, max_queue_len, queue_len_factor)
+    mem = fifo.seq_mem_d1("mem", val_width, max_queue_len, queue_len_factor)
     write = fifo.reg(queue_len_factor)  # The next address to write to
     read = fifo.reg(queue_len_factor)  # The next address to read from
     # We will orchestrate `mem`, along with the two pointers above, to
     # simulate a circular queue of size 2^queue_len_factor.
 
-    ans = fifo.reg(32, "ans", is_ref=True)
+    ans = fifo.reg(val_width, "ans", is_ref=True)
     # If the user wants to pop or peek, we will write the value to `ans`.
     err = fifo.reg(1, "err", is_ref=True)
     # We'll raise this as a general error flag for overflow and underflow.
-    len = fifo.reg(32)  # The active length of the FIFO.
+    len = fifo.reg(bits_needed(max_queue_len))  # The active length of the FIFO.
```

Comment:
@polybeandip I wonder if something like this could be the answer to your length-calculation question in https://github.com/calyxir/calyx/pull/2164. My `len` used to be a 32-bit number, which was obviously gratuitous. Moving it to `queue_len_factor` is the obvious choice, but then I'd have the issue where `len` would never actually equal `max_queue_len = 2**queue_len_factor` because of overflow. 

This little `bits_needed` thing is a method that ensures that `len` _can_ equal `max_queue_len`. Of course, you and I know that `bits_needed(max_queue_len)` will always equal `queue_len_factor + 1`. There is no problem later when it comes to things like `fifo.eq_use(len.out, 0)` and `fifo.eq_use(len.out, max_queue_len)` because `0` and `max_queue_len` get to _just be Python numbers_ which later benefit from eDSL width-inference.

Reply:
☝️ is just to explore the space. In reality I think your `is_full` flag is totally solid and could just be left as-is.